### PR TITLE
[issue-321] introduce dataclass (with properties) and typeguard

### PR DESCRIPTION
**Pull request by armintaenzertng**
Fixes https://github.com/spdx/tools-python/issues/321

Signed-off-by: Armin Tänzer <armin.taenzer@tngtech.com>

### CODE REVIEW

File: tests/model/test_external_package_reference.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+                                                          "comment")
+    assert external_package_reference.category == ExternalPackageReferenceCategory.OTHER
+    assert external_package_reference.reference_type == "type"
+    assert external_package_reference.locator == "locator"
+    assert external_package_reference.comment == "comment"
+
+
+def test_wrong_type_in_category():
+    with pytest.raises(TypeError):
+        ExternalPackageReference([ExternalPackageReferenceCategory.OTHER], 42, "locator", "comment")
```
Comment (nicoweidner) [inline tests/model/test_external_package_reference.py:17]: nit: Should have a valid type

**Review by nicoweidner — COMMENTED** [review]
Really cool changes! I like that we can get rid of the boilerplate constructors this way. The most important suggestion would be to combine the three annotations that we add on every data model class into one annotation

File: src/model/dataclass_with_properties.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+from typing import Type, Any
+
+
+def dataclass_with_properties(cls):
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply src/model/dataclass_with_properties.py:4]: implemented :)

File: src/model/dataclass_with_properties.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+def make_getter(field_name, field_type):
+    """helper method to avoid late binding when generating functions in a for loop"""
+    def get_field(self) -> field_type:
+        return getattr(self, f"_{field_name}")
+
+    return get_field
+
+
+def get_value_error_message(field_name: str, expected_type: Type, value: Any):
+    return f"{field_name} must be {expected_type} but is {type(value)} instead: {value}"
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply src/model/dataclass_with_properties.py:37]: removed

File: src/model/license_expression.py [hunk 4]
```diff
@@ -8,8 +8,10 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+from dataclasses import dataclass
 
 
+@dataclass(eq=True)
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply src/model/license_expression.py:14]: replaced by the new decorator

File: src/model/package.py [hunk 5]
```diff
@@ -36,14 +38,13 @@ class PackagePurpose(Enum):
     OTHER = auto()
 
 
+@typechecked
+@dataclass_with_properties
+@dataclass(eq=True)
 class PackageVerificationCode:
     value: str
     excluded_files: List[str]
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply src/model/package.py:46]: fixed

File: src/model/package.py [hunk 6]
```diff
@@ -61,80 +65,34 @@ class ExternalPackageReference:
     locator: str
     comment: Optional[str]
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply src/model/package.py:66]: fixed

File: tests/model/test_external_package_reference.py (hunk 1 above)
Comment (armintaenzertng) [reply tests/model/test_external_package_reference.py:17]: fixed

File: tests/model/test_relationship.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    relationship = Relationship("id", RelationshipType.OTHER, "other_id", "comment")
+    assert relationship.spdx_element_id == "id"
+    assert relationship.relationship_type == RelationshipType.OTHER
+    assert relationship.related_spdx_element_id == "other_id"
+    assert relationship.comment == "comment"
+
+
+def test_wrong_type_in_spdx_element_id():
+    with pytest.raises(TypeError):
+        Relationship(42, RelationshipType.OTHER, "other_id", "comment")
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply tests/model/test_relationship.py:16]: fixed

File: tests/model/test_document.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
+from unittest import mock
+
+import pytest
+
+from src.model.document import Document
+
+
+@mock.patch('src.model.document.CreationInfo', autospec=True)
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply tests/model/test_document.py:8]: I'd rather leave that for later

File: src/model/dataclass_with_properties.py [hunk 9]
```diff
@@ -1,24 +1,28 @@
...
+    data_cls = dataclass(cls)
+    for field_name, field_type in data_cls.__annotations__.items():
 
         set_field = make_setter(field_name, field_type)
         get_field = make_getter(field_name, field_type)
 
-        setattr(cls, f"set_{field_name}", set_field)
-        setattr(cls, f"get_{field_name}", get_field)
+        setattr(data_cls, f"set_{field_name}", set_field)
+        setattr(data_cls, f"get_{field_name}", get_field)
```
Comment (nicoweidner) [inline src/model/dataclass_with_properties.py:16]: nit: I'm pretty sure I verified that you don't actually need to set the `set_x` and `get_x` methods on the class. You can simply use the getter and setter that were defined above directly. Doesn't really matter either way, though.

File: src/model/dataclass_with_properties.py [hunk 10]
```diff
@@ -1,24 +1,28 @@
+from dataclasses import dataclass
 from typing import Type, Any
```
Comment (nicoweidner) [inline src/model/dataclass_with_properties.py:2]: nit: unused imports

**Review by nicoweidner — APPROVED** [review]
Found two more nits, but this looks really good now!

**Pull request merged**